src/searchers/astar.rs

```diff
@@ -30,16 +30,159 @@ use crate::filtration_system::{
+/// - "atbash" maps to "AtbashDecoder"
+/// - "caesar" maps to "CaesarDecoder"
+static CIPHER_MAPPING: Lazy<HashMap<&'static str, &'static str>> = Lazy::new(|| {
+    let mut map = HashMap::new();
+    map.insert("fractionatedMorse", "MorseCodeDecoder");
+    map.insert("atbash", "AtbashDecoder");
+    map.insert("caesar", "CaesarDecoder");
+    map.insert("railfence", "RailfenceDecoder");
+    map.insert("rot47", "ROT47Decoder");
+    map.insert("a1z26", "A1Z26Decoder");
+    map.insert("simplesubstitution", "SubstitutionGenericDecoder");
+    // Add more mappings as needed
+    map
+});
+
+/// Track decoder success rates for adaptive learning
+static DECODER_SUCCESS_RATES: Lazy<Mutex<HashMap<String, (usize, usize)>>> =
+    Lazy::new(|| Mutex::new(HashMap::new()));
+
+/// Update decoder statistics based on success or failure
+///
+/// # Arguments
+///
+/// * `decoder` - The name of the decoder
+/// * `success` - Whether the decoder was successful
+fn update_decoder_stats(decoder: &str, success: bool) {
+    let mut stats = DECODER_SUCCESS_RATES.lock().unwrap();
+    let (successes, total) = stats.entry(decoder.to_string()).or_insert((0, 0));
+
+    if success {
+        *successes += 1;
+    }
+    *total += 1;
+
+    // TODO: Write this data to a file for persistence
+}
+
+/// Get the success rate of a decoder
+///
+/// # Arguments
+///
+/// * `decoder` - The name of the decoder
+///
+/// # Returns
+///
+/// * The success rate as a float between 0.0 and 1.0
+fn get_decoder_success_rate(decoder: &str) -> f32 {
+    let stats = DECODER_SUCCESS_RATES.lock().unwrap();
+    if let Some((successes, total)) = stats.get(decoder) {
+        if *total > 0 {
+            return *successes as f32 / *total as f32;
+        }
+    }
+
+    // Default for unknown decoders
+    0.5
+}
+
+/// Get the cipher identification score for a text
+///
+/// # Arguments
+///
+/// * `text` - The text to analyze
+///
+/// # Returns
+///
+/// * A tuple containing the identified cipher and its score
+fn get_cipher_identifier_score(text: &str) -> (String, f32) {
+    let results = cipher_identifier::identify_cipher::identify_cipher(text, 5, None);
+
+    for (cipher, score) in results {
+        if let Some(_decoder) = CIPHER_MAPPING.get(cipher.as_str()) {
+            return (cipher, (score / 10.0) as f32);
+        }
+    }
+
+    // Default if no match
+    let mut rng = rand::thread_rng();
+    ("unknown".to_string(), rng.gen_range(0.5..1.0) as f32)
+}
+
+/// Check if a decoder and cipher form a common sequence
+///
+/// # Arguments
+///
+/// * `prev_decoder` - The name of the previous decoder
+/// * `current_cipher` - The name of the current cipher
+///
+/// # Returns
+///
+/// * `true` if the sequence is common, `false` otherwise
+fn is_common_sequence(prev_decoder: &str, current_cipher: &str) -> bool {
+    // Define common sequences of decoders
+    match (prev_decoder, current_cipher) {
+        ("Base64Decoder", "caesar") => true,
```

Comment:
caesar -> caesar bad. binary -> binary bad. hex -> hex might also be bad here. I think most base decoders are good here